consensus/polybft/consensus_runtime.go

```diff
@@ -483,73 +531,128 @@ func (c *consensusRuntime) restartEpoch(header *types.Header) (*epochMetadata, e
 		return nil, err
 	}
 
+	if err := c.governanceManager.PostEpoch(reqObj); err != nil {
+		return nil, err
+	}
+
+	currentConfig, err := c.governanceManager.GetClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
 	return &epochMetadata{
-		Number:            epochNumber,
-		Validators:        validatorSet,
-		FirstBlockInEpoch: firstBlockInEpoch,
+		Number:              epochNumber,
+		Validators:          validatorSet,
+		FirstBlockInEpoch:   firstBlockInEpoch,
+		CurrentClientConfig: currentConfig,
 	}, nil
 }
 
-// calculateCommitEpochInput calculates commit epoch input data for blocks starting from the last built block
-// in the current epoch, and ending at the last block of previous epoch
-func (c *consensusRuntime) calculateCommitEpochInput(
-	currentBlock *types.Header,
-	epoch *epochMetadata,
-) (*contractsapi.CommitEpochValidatorSetFn,
-	*contractsapi.DistributeRewardForRewardPoolFn, error) {
-	uptimeCounter := map[types.Address]int64{}
-	blockHeader := currentBlock
-	epochID := epoch.Number
-	totalBlocks := int64(0)
+// createCommitEpochInput creates commit epoch input data
+func createCommitEpochInput(
+	currentBlock *types.Header, epoch *epochMetadata) *contractsapi.CommitEpochValidatorSetFn {
+	return &contractsapi.CommitEpochValidatorSetFn{
+		ID: new(big.Int).SetUint64(epoch.Number),
+		Epoch: &contractsapi.Epoch{
+			StartBlock: new(big.Int).SetUint64(epoch.FirstBlockInEpoch),
+			EndBlock:   new(big.Int).SetUint64(currentBlock.Number + 1),
+			EpochRoot:  types.Hash{},
+		},
+		EpochSize: new(big.Int).SetUint64(epoch.CurrentClientConfig.EpochSize),
+	}
+}
+
+// calculateDistributeRewardsInput calculates distribute rewards input data
+func (c *consensusRuntime) calculateDistributeRewardsInput(
+	isFirstBlockOfEpoch, isEndOfEpoch bool,
+	pendingBlockNumber uint64,
+	lastFinalizedBlock *types.Header,
+	epochID uint64,
+) (*contractsapi.DistributeRewardForRewardPoolFn, error) {
+	if !isRewardDistributionBlock(isFirstBlockOfEpoch, isEndOfEpoch, pendingBlockNumber) {
+		// we don't have to distribute rewards at this block
+		return nil, nil
+	}
+
+	var (
+		// epoch size is the number of blocks that really happened
+		// because of slashing, epochs might not have the configured number of blocks
+		epochSize     = uint64(0)
+		uptimeCounter = map[types.Address]int64{}
+		blockHeader   = lastFinalizedBlock // start calculating from this block
+	)
+
+	if forkmanager.GetInstance().IsForkEnabled(chain.Governance, pendingBlockNumber) {
+		// if governance is enabled, we are distributing rewards for previous epoch
+		// at the beginning of a new epoch, so modify epochID
+		epochID--
```

Comment:
WHat is the current epoch ID is 0?

Reply:
Current epoch can not be 0. We read epoch ID from ValidatorSet contract, and it's starting value is always 1.